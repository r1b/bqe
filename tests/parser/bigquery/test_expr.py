import pytest


@pytest.mark.parametrize("sql", (
    pytest.param("SELECT column1.column2", id="dotted_column_expr_single"),
    pytest.param("SELECT column1.column2.column3", id="dotted_column_expr_multi"),
    pytest.param("SELECT column1 FROM dataset1.table1", id="dotted_table_expr_single"),
    pytest.param("SELECT column1 FROM `project1`.dataset1.table1", id="dotted_table_expr_multi"),
))
def test_dotted_expr_ok(sql, assert_parse_tree):
    assert_parse_tree(sql)


@pytest.mark.parametrize("sql", (
    pytest.param("SELECT column1[0]", id="subscript_expr_integer_literal_single"),
    pytest.param("SELECT column1[0][1]", id="subscript_expr_integer_literal_multi"),
    pytest.param("SELECT column1.column2[0][1]", id="subscript_expr_path_expr_nested"),
    pytest.param("SELECT column1[OFFSET(2)]", id="subscript_expr_offset"),
    pytest.param("SELECT column1['key1']", id="subscript_expr_string_literal"),
))
def test_subscript_expr_ok(sql, assert_parse_tree):
    assert_parse_tree(sql)


@pytest.mark.parametrize("sql", (
    pytest.param("SELECT FUNC()", id="fcall_pos_arg_none"),
    pytest.param("SELECT FUNC(1)", id="fcall_pos_arg_single"),
    pytest.param("SELECT FUNC(1, 2)", id="fcall_pos_arg_multi"),
    pytest.param("SELECT FUNC(name1 => 1)", id="fcall_named_arg_single"),
    pytest.param("SELECT FUNC(name1 => 1, name2 => 2)", id="fcall_named_arg_multi"),
    pytest.param("SELECT FUNC(1, name2 => 2)", id="fcall_kitchen_sink"),
))
def test_function_call_ok(sql, assert_parse_tree):
    assert_parse_tree(sql)


@pytest.mark.parametrize("sql", (
    pytest.param("SELECT FUNC(", id="fcall_unclosed_paren"),
    pytest.param("SELECT FUNC(,)", id="fcall_comma_only"),
    pytest.param("SELECT FUNC(,1)", id="fcall_pos_comma_leading"),
    pytest.param("SELECT FUNC(1,)", id="fcall_pos_comma_trailing"),
    pytest.param("SELECT FUNC(,name1 => 1)", id="fcall_named_comma_leading"),
    pytest.param("SELECT FUNC(name1 => 1,)", id="fcall_named_comma_trailing"),
    # TODO: Explicit error for this one
    pytest.param("SELECT FUNC(name1 => 1, 2)", id="fcall_named_before_pos"),
))
def test_function_call_error(sql, assert_parse_tree_error):
    assert_parse_tree_error(sql)
