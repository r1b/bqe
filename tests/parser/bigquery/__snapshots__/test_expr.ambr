# serializer version: 1
# name: test_dotted_expr_ok[dotted_column_expr_multi]
  '''
  start
    query_statement
      query_expr
        select
          select_list
            bigquery__select_item
              bigquery__select_expr
                bigquery__column_expr
                  bigquery__column_expr
                    bigquery__column_expr
                      bigquery__ident	column1
                    bigquery__ident	column2
                  bigquery__ident	column3
  
  '''
# ---
# name: test_dotted_expr_ok[dotted_column_expr_single]
  '''
  start
    query_statement
      query_expr
        select
          select_list
            bigquery__select_item
              bigquery__select_expr
                bigquery__column_expr
                  bigquery__column_expr
                    bigquery__ident	column1
                  bigquery__ident	column2
  
  '''
# ---
# name: test_dotted_expr_ok[dotted_table_expr_multi]
  '''
  start
    query_statement
      query_expr
        select
          select_list
            bigquery__select_item
              bigquery__select_expr
                bigquery__column_expr
                  bigquery__ident	column1
          bigquery__select_from
            from_clause
              bigquery__from_item_table
                bigquery__path_expression
                  bigquery__path_expression
                    bigquery__path_expression
                      bigquery__ident	`project1`
                    bigquery__ident	dataset1
                  bigquery__ident	table1
  
  '''
# ---
# name: test_dotted_expr_ok[dotted_table_expr_single]
  '''
  start
    query_statement
      query_expr
        select
          select_list
            bigquery__select_item
              bigquery__select_expr
                bigquery__column_expr
                  bigquery__ident	column1
          bigquery__select_from
            from_clause
              bigquery__from_item_table
                bigquery__path_expression
                  bigquery__path_expression
                    bigquery__ident	dataset1
                  bigquery__ident	table1
  
  '''
# ---
# name: test_function_call_ok[fcall_kitchen_sink]
  '''
  start
    query_statement
      query_expr
        select
          select_list
            bigquery__select_item
              bigquery__select_expr
                bigquery__column_expr
                  bigquery__function_call_expr
                    bigquery__column_expr
                      bigquery__ident	FUNC
                    bigquery__function_args
                      bigquery__pos_arg
                        bigquery__column_expr
                          bigquery__literal
                            bigquery__integer_literal	1
                      bigquery__named_arg
                        bigquery__ident	name2
                        bigquery__column_expr
                          bigquery__literal
                            bigquery__integer_literal	2
  
  '''
# ---
# name: test_function_call_ok[fcall_named_arg_multi]
  '''
  start
    query_statement
      query_expr
        select
          select_list
            bigquery__select_item
              bigquery__select_expr
                bigquery__column_expr
                  bigquery__function_call_expr
                    bigquery__column_expr
                      bigquery__ident	FUNC
                    bigquery__function_args
                      bigquery__named_arg
                        bigquery__ident	name1
                        bigquery__column_expr
                          bigquery__literal
                            bigquery__integer_literal	1
                      bigquery__named_arg
                        bigquery__ident	name2
                        bigquery__column_expr
                          bigquery__literal
                            bigquery__integer_literal	2
  
  '''
# ---
# name: test_function_call_ok[fcall_named_arg_single]
  '''
  start
    query_statement
      query_expr
        select
          select_list
            bigquery__select_item
              bigquery__select_expr
                bigquery__column_expr
                  bigquery__function_call_expr
                    bigquery__column_expr
                      bigquery__ident	FUNC
                    bigquery__function_args
                      bigquery__named_arg
                        bigquery__ident	name1
                        bigquery__column_expr
                          bigquery__literal
                            bigquery__integer_literal	1
  
  '''
# ---
# name: test_function_call_ok[fcall_pos_arg_multi]
  '''
  start
    query_statement
      query_expr
        select
          select_list
            bigquery__select_item
              bigquery__select_expr
                bigquery__column_expr
                  bigquery__function_call_expr
                    bigquery__column_expr
                      bigquery__ident	FUNC
                    bigquery__function_args
                      bigquery__pos_arg
                        bigquery__column_expr
                          bigquery__literal
                            bigquery__integer_literal	1
                      bigquery__pos_arg
                        bigquery__column_expr
                          bigquery__literal
                            bigquery__integer_literal	2
  
  '''
# ---
# name: test_function_call_ok[fcall_pos_arg_none]
  '''
  start
    query_statement
      query_expr
        select
          select_list
            bigquery__select_item
              bigquery__select_expr
                bigquery__column_expr
                  bigquery__function_call_expr
                    bigquery__column_expr
                      bigquery__ident	FUNC
  
  '''
# ---
# name: test_function_call_ok[fcall_pos_arg_single]
  '''
  start
    query_statement
      query_expr
        select
          select_list
            bigquery__select_item
              bigquery__select_expr
                bigquery__column_expr
                  bigquery__function_call_expr
                    bigquery__column_expr
                      bigquery__ident	FUNC
                    bigquery__function_args
                      bigquery__pos_arg
                        bigquery__column_expr
                          bigquery__literal
                            bigquery__integer_literal	1
  
  '''
# ---
# name: test_subscript_expr_ok[subscript_expr_integer_literal_multi]
  '''
  start
    query_statement
      query_expr
        select
          select_list
            bigquery__select_item
              bigquery__select_expr
                bigquery__column_expr
                  bigquery__subscript_expr
                    bigquery__column_expr
                      bigquery__subscript_expr
                        bigquery__column_expr
                          bigquery__ident	column1
                        bigquery__column_expr
                          bigquery__literal
                            bigquery__integer_literal	0
                    bigquery__column_expr
                      bigquery__literal
                        bigquery__integer_literal	1
  
  '''
# ---
# name: test_subscript_expr_ok[subscript_expr_integer_literal_single]
  '''
  start
    query_statement
      query_expr
        select
          select_list
            bigquery__select_item
              bigquery__select_expr
                bigquery__column_expr
                  bigquery__subscript_expr
                    bigquery__column_expr
                      bigquery__ident	column1
                    bigquery__column_expr
                      bigquery__literal
                        bigquery__integer_literal	0
  
  '''
# ---
# name: test_subscript_expr_ok[subscript_expr_offset]
  '''
  start
    query_statement
      query_expr
        select
          select_list
            bigquery__select_item
              bigquery__select_expr
                bigquery__column_expr
                  bigquery__subscript_expr
                    bigquery__column_expr
                      bigquery__ident	column1
                    bigquery__column_expr
                      bigquery__function_call_expr
                        bigquery__column_expr
                          bigquery__ident	OFFSET
                        bigquery__function_args
                          bigquery__pos_arg
                            bigquery__column_expr
                              bigquery__literal
                                bigquery__integer_literal	2
  
  '''
# ---
# name: test_subscript_expr_ok[subscript_expr_path_expr_nested]
  '''
  start
    query_statement
      query_expr
        select
          select_list
            bigquery__select_item
              bigquery__select_expr
                bigquery__column_expr
                  bigquery__subscript_expr
                    bigquery__column_expr
                      bigquery__subscript_expr
                        bigquery__column_expr
                          bigquery__column_expr
                            bigquery__ident	column1
                          bigquery__ident	column2
                        bigquery__column_expr
                          bigquery__literal
                            bigquery__integer_literal	0
                    bigquery__column_expr
                      bigquery__literal
                        bigquery__integer_literal	1
  
  '''
# ---
# name: test_subscript_expr_ok[subscript_expr_string_literal]
  '''
  start
    query_statement
      query_expr
        select
          select_list
            bigquery__select_item
              bigquery__select_expr
                bigquery__column_expr
                  bigquery__subscript_expr
                    bigquery__column_expr
                      bigquery__ident	column1
                    bigquery__column_expr
                      bigquery__literal
                        bigquery__string_literal	'key1'
  
  '''
# ---
